File: leetcode/0105-construct-binary-tree-from-preorder-and-inorder-traversal/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        return buildTree(preorder, 0, preorder.size()-1, inorder, 0, inorder.size()-1);
    }
    
    TreeNode* buildTree(vector<int>& preorder, int pStart, int pEnd, vector<int>& inorder, int iStart, int iEnd) {
        if (pEnd < pStart)
            return nullptr;
        
        if (pStart == pEnd)
            return new TreeNode(preorder[pStart]);
        
        TreeNode* root = new TreeNode(preorder[pStart]);
        auto iterIndexOfRootInorder = find(inorder.begin() + iStart, inorder.begin() + iEnd+1, preorder[pStart]);
        int indexOfRootInorder = iterIndexOfRootInorder - inorder.begin();
        int numOfNodesInLeftSubTree = indexOfRootInorder - iStart;
        int numOfNodesInRightSubTree = iEnd - indexOfRootInorder;
        root->left = buildTree(preorder, pStart+1, pStart + numOfNodesInLeftSubTree,
                               inorder, iStart, indexOfRootInorder - 1);
        root->right = buildTree(preorder, pStart + numOfNodesInLeftSubTree + 1, pEnd,
                                inorder, indexOfRootInorder + 1, iEnd);
        
        return root;
    }
};

File: leetcode/0105-construct-binary-tree-from-preorder-and-inorder-traversal/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        return self.buildTreeUtil(preorder, 0, len(preorder) - 1, inorder, 0, len(inorder) - 1)
    
    def buildTreeUtil(self, preorder :List[int], pStart :int, pEnd: int, inorder: List[int], iStart: int, iEnd: int) -> TreeNode:
        if pEnd < pStart:
            return None
        
        if pStart == pEnd:
            return TreeNode(preorder[pStart])
        
        root = TreeNode(preorder[pStart])
        indexOfRootInorder = inorder.index(root.val, iStart)
        
        numOfNodesInLeftSubTree = indexOfRootInorder - iStart
        numOfNodesInRightSubTree = iEnd - indexOfRootInorder
        root.left = self.buildTreeUtil(preorder, pStart+1, pStart + numOfNodesInLeftSubTree,
                               inorder, iStart, indexOfRootInorder - 1)
        root.right = self.buildTreeUtil(preorder, pStart + numOfNodesInLeftSubTree + 1, pEnd,
                                inorder, indexOfRootInorder + 1, iEnd)
        
        return root